src/main/kotlin/org/rust/ide/template/postfix/stringBasedPostfixTemplates.kt

```diff
@@ -0,0 +1,76 @@
+package org.rust.ide.template.postfix
+
+import com.intellij.codeInsight.template.Template
+import com.intellij.codeInsight.template.impl.TextExpression
+import com.intellij.codeInsight.template.postfix.templates.StringBasedPostfixTemplate
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RustBinaryExprElement
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.RustTupleFieldDeclElement
+import org.rust.lang.core.psi.impl.RustPsiImplUtil
+import org.rust.lang.core.psi.util.descendentsOfType
+import org.rust.lang.core.psi.util.parentOfType
+import org.rust.lang.core.resolve.RustResolveEngine
+import org.rust.lang.core.resolve.scope.RustResolveScope
+import org.rust.lang.core.types.RustEnumType
+import org.rust.lang.core.types.util.resolvedType
+
+class MatchPostfixTemplate : StringBasedPostfixTemplate(
+    "match",
+    "match expr {...}",
+    RustTopMostInScopeSelector(RustExprElement::isEnum)
+) {
+    override fun getTemplateString(element: PsiElement): String? {
+        val enumType = (element as RustExprElement).resolvedType as RustEnumType
+
+        val allAvailableNames = generateSequence(
+            element.parentOfType<RustResolveScope>(),
+            { it.parentOfType<RustResolveScope>() }
+        )
+            .flatMap { RustResolveEngine.declarations(it, element) }
+            .mapNotNull { it.element }
+            .mapNotNull { RustPsiImplUtil.canonicalCratePath(it) }
+            .map { it.toString() }
+            .toSet()
+
+        val stringBuilder = StringBuilder()
+        stringBuilder.append("match ${element.text} {\n")
+
+        for ((i, item) in enumType.item.enumBody.enumVariantList.withIndex()) {
+            val itemCanonicalName = RustPsiImplUtil.canonicalCratePath(item).toString()
+
+            val excessPath = allAvailableNames
```

Comment:
Yes, it's costly. But every enum variant name checks for cases like this:
```Rust
        enum MyOption<T> {
            None,
            Some(T)
        }
        fn main() {
            let a = MyOption::Some(42);
            use MyOption::None;
            match a {
                None => {},
                MyOption::Some(v0) => {},
            };
        }
```
Could you specify where there may be a problem with correctness?
1. We have enum variant. We can get the full path to it.
2. We can get all declarations, which available from the current scope.
3. We can get get absolute path for every declaration.
4. If some declaration from enum-variant path present in the current scope, when not necessary to specify full path to it. And we can drop some prefix of the path.